gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +273,439 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
 
 
 /* ************************************************************************* */
-
 /**
- * A convenient base class for creating your own NoiseModelFactor with 1
- * variable.  To derive from this class, implement evaluateError().
+ * A convenient base class for creating your own NoiseModelFactor
+ * with n variables.  To derive from this class, implement evaluateError().
+ *
+ * For example, a 2-way factor could be implemented like so:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * class MyFactor : public NoiseModelFactorN<double, double> {
+ *  public:
+ *   using Base = NoiseModelFactorN<double, double>;
+ *
+ *   MyFactor(Key key1, Key key2, const SharedNoiseModel& noiseModel)
+ *       : Base(noiseModel, key1, key2) {}
  *
- * Templated on a values structure type. The values structures are typically
- * more general than just vectors, e.g., Rot3 or Pose3,
- * which are objects in non-linear manifolds (Lie groups).
+ *   Vector evaluateError(
+ *       const double& x1, const double& x2,
+ *       boost::optional<Matrix&> H1 = boost::none,
+ *       boost::optional<Matrix&> H2 = boost::none) const override {
+ *     if (H1) *H1 = (Matrix(1, 1) << 1.0).finished();
+ *     if (H2) *H2 = (Matrix(1, 1) << 2.0).finished();
+ *     return (Vector(1) << x1 + 2 * x2).finished();
+ *   }
+ * };
+ * ~~~~~~~~~~~~~~~~~~~~
+ *
+ * These factors are templated on a values structure type. The values structures
+ * are typically more general than just vectors, e.g., Rot3 or Pose3, which are
+ * objects in non-linear manifolds (Lie groups).
  */
-template<class VALUE>
-class NoiseModelFactor1: public NoiseModelFactor {
-
-public:
-
-  // typedefs for value types pulled from keys
-  typedef VALUE X;
-
-protected:
-
-  typedef NoiseModelFactor Base;
-  typedef NoiseModelFactor1<VALUE> This;
-
-public:
+template <class... VALUES>
+class NoiseModelFactorN : public NoiseModelFactor {
+ public:
+  /// N is the number of variables (N-way factor)
+  enum { N = sizeof...(VALUES) };
+
+  /** The type of the i'th template param can be obtained as X<I> */
+  template <int I, typename std::enable_if<(I < N), bool>::type = true>
+  using X = typename std::tuple_element<I, std::tuple<VALUES...>>::type;
```

Comment:
Would you be opposed to renaming this from `X` to `TYPE` or `T`? I can envision a scenario where a user sees `X` and thinks it is the symbol shorthand.

From a readability perspective, I prefer `MyFactor::TYPE<1>`.

Reply:
Or even `MyFactor::type<1>` similar to `MyFactor::key<1>` later.

Reply:
I'm happy to change, e.g. to `T<#>`, but my logic here was to match the old API, which uses `X1`, `X2`, ...
I think this was probably done in the original version since the input variable (e.g. in unwhitenedError, evaluateError, etc.) is always named with `x`, so the type should be capital `X`.  I guess `x` is used to denote an "input" variable, as in f(x).

Can you please re-confirm whether you prefer this change?  Then I'll apply it or resolve convo.

Reply:
If the old API had `X` then let's match it.

Reply:
I noticed the mucking in the tests with avoiding collisions with symbol_shorthand::X, so I think we should revisit. I propose to rename

`VALUES -> ValueTypes` 

```
using ValueType = ...
```

Reply:
Done.